src/eins-hwinfo.c

```diff
@@ -0,0 +1,229 @@
+                                                 NULL, NULL);
+
+  return info != NULL &&
+    g_file_info_get_attribute_boolean (info, G_FILE_ATTRIBUTE_UNIX_IS_MOUNTPOINT);
+}
+
+/**
+ * @data: G_SOURCE_REMOVE or G_SOURCE_CONTINUE
+ */
+static gboolean
+record_disk_space (gpointer data)
+{
+  g_autoptr(GFile) root = g_file_new_for_path ("/");
+  g_autoptr(GFile) extra = g_file_new_for_path ("/var/endless-extra");
+  GVariant *root_payload = get_disk_space_for_partition (root);
+  GVariant *extra_payload = is_mounted (extra)
+    ? get_disk_space_for_partition (extra)
+    : g_variant_new ("(ttt)", 0, 0, 0);
+  GVariant *payload = g_variant_new ("(@(ttt)@(ttt))", root_payload, extra_payload);
+
+  emtr_event_recorder_record_event (emtr_event_recorder_get_default (),
+                                    DISK_SPACE_EVENT,
+                                    g_steal_pointer (&payload));
+
+  return GPOINTER_TO_INT (data);
+}
+
+static gboolean
+record_ram_size (gpointer unused)
+{
+  glibtop_mem mem;
+  GVariant *payload;
+
+  glibtop_get_mem (&mem);
+  payload = g_variant_new_uint64 (mem.total);
+  emtr_event_recorder_record_event (emtr_event_recorder_get_default (),
+                                    RAM_SIZE_EVENT,
+                                    g_steal_pointer (&payload));
+  return G_SOURCE_REMOVE;
+}
+
+/* Derived from gnome-control-center's get_cpu_info(). */
+static GVariant *
+get_cpu_info (const glibtop_sysinfo *info)
+{
+  /* Keys: gchar *, borrowed from info
+   * Values: guint32
+   */
+  g_autoptr(GHashTable) counts = g_hash_table_new (g_str_hash, g_str_equal);
+  GHashTableIter iter;
+  gpointer key, value;
+  int i, j;
+  GVariantBuilder builder;
+
+  /* count duplicates */
+  for (i = 0; i != info->ncpu; ++i)
+    {
+      const char * const keys[] = { "model name", "cpu", "Processor" };
+      char *model;
+      int  *count;
+
+      model = NULL;
+
+      for (j = 0; model == NULL && j != G_N_ELEMENTS (keys); ++j)
+        {
+          model = g_hash_table_lookup (info->cpuinfo[i].values,
+                                       keys[j]);
+        }
+
+      if (model == NULL)
+        continue;
+
+      count = g_hash_table_lookup (counts, model);
+      if (count == NULL)
+        g_hash_table_insert (counts, model, GUINT_TO_POINTER (1));
+      else
+        g_hash_table_replace (counts, model, GUINT_TO_POINTER (GPOINTER_TO_UINT (count) + 1));
+    }
+
+  g_variant_builder_init (&builder, G_VARIANT_TYPE_ARRAY);
+  g_hash_table_iter_init (&iter, counts);
+  while (g_hash_table_iter_next (&iter, &key, &value))
+    {
+      const char *model = key;
+      guint count = GPOINTER_TO_INT (value);
+
+      if (count > G_MAXUINT16)
+        {
+          g_warning ("%s has %u threads; clamping to %" G_GUINT16_FORMAT,
+                     model, count, G_MAXUINT16);
+          count = G_MAXUINT16;
+        }
+
+      g_variant_builder_add (&builder, "{sq}", model, count);
```

Comment:
This assumes that `model` is valid UTF-8. The [Wikipedia page for CPUID on x86](https://en.wikipedia.org/wiki/CPUID#EAX=80000002h,80000003h,80000004h:_Processor_Brand_String) says that the processor brand string is ASCII (a subset of UTF-8); and the implementation for ARM is ASCII by construction in the kernel.

However, on ARM 'model name' is just something like "ARMv7 Processor rev 0 (v7l)" which doesn't really tell us what we need…

Reply:
Addressed by reimplementing this in terms of `lscpu` which decodes the additional fields necessary to get a more useful model name on ARM.